File: string/Trie/cpp/Trie.cpp
#include "Trie.hpp"

#include <stdexcept>

using namespace std;

Trie::Trie() : root(make_unique<Node>()) {}

void Trie::insert(const string& word) {
    Node* current = root.get();
    for (char ch : word) {
        const int index = ch - 'a';
        if (index < 0 || index >= 26) {
            throw invalid_argument("Trie only supports lowercase English letters");
        }
        if (!current->children[index]) {
            current->children[index] = make_unique<Node>();
        }
        current = current->children[index].get();
    }
    current->is_word = true;
}

bool Trie::search(const string& word) const {
    const Node* node = findNode(word);
    return node != nullptr && node->is_word;
}

bool Trie::startsWith(const string& prefix) const {
    return findNode(prefix) != nullptr;
}

const Trie::Node* Trie::findNode(const string& text) const {
    const Node* current = root.get();
    for (char ch : text) {
        const int index = ch - 'a';
        if (index < 0 || index >= 26) {
            throw invalid_argument("Trie only supports lowercase English letters");
        }
        if (!current->children[index]) {
            return nullptr;
        }
        current = current->children[index].get();
    }
    return current;
}

vector<string> executeTrieCommands(const vector<TrieCommand>& commands) {
    Trie trie;
    vector<string> output;

    for (const TrieCommand& command : commands) {
        if (command.operation == "insert") {
            trie.insert(command.value);
        } else if (command.operation == "search") {
            output.push_back(trie.search(command.value) ? "true" : "false");
        } else if (command.operation == "starts_with") {
            output.push_back(trie.startsWith(command.value) ? "true" : "false");
        } else {
            throw invalid_argument("Unknown trie command: " + command.operation);
        }
    }

    return output;
}

File: string/Trie/python/Trie.py
from __future__ import annotations

from dataclasses import dataclass, field


@dataclass(slots=True)
class _Node:
    children: list[_Node | None] = field(default_factory=lambda: [None] * 26)
    is_word: bool = False


class Trie:
    def __init__(self) -> None:
        self._root = _Node()

    def insert(self, word: str) -> None:
        current = self._root
        for char in word:
            index = self._index(char)
            if current.children[index] is None:
                current.children[index] = _Node()
            current = current.children[index]
        current.is_word = True

    def search(self, word: str) -> bool:
        node = self._find_node(word)
        return node is not None and node.is_word

    def starts_with(self, prefix: str) -> bool:
        return self._find_node(prefix) is not None

    def _find_node(self, text: str) -> _Node | None:
        current = self._root
        for char in text:
            index = self._index(char)
            child = current.children[index]
            if child is None:
                return None
            current = child
        return current

    @staticmethod
    def _index(char: str) -> int:
        index = ord(char) - ord("a")
        if index < 0 or index >= 26:
            raise ValueError("Trie only supports lowercase English letters")
        return index


def execute_trie_commands(commands: list[tuple[str, str]]) -> list[str]:
    trie = Trie()
    output: list[str] = []

    for operation, value in commands:
        if operation == "insert":
            trie.insert(value)
        elif operation == "search":
            output.append("true" if trie.search(value) else "false")
        elif operation == "starts_with":
            output.append("true" if trie.starts_with(value) else "false")
        else:
            raise ValueError(f"Unknown trie command: {operation}")

    return output
